### ISSUE TITLE
Hang (deadlock?) without error message during analyzer failure

### ISSUE THREAD

**Issue by philrz** [brimdata/brimcap#71]
As part of my work on #64, I've created the following wrapper script to invoke a custom Suricata install, which uses `jq` to work around brimdata/zed#2523:

```
$ cat suricata-wrapper.sh 
#!/bin/bash
suricata -r /dev/stdin
cat eve.json | jq -c . > deduped-eve.json
```

I invoke this via Brimcap config YAML:

```
$ cat suricata.yml 
analyzers:
  - cmd: suricata-wrapper.sh
    globs: ["deduped*.json"]
    shaper: |
      type alert = {
        timestamp: time,
        event_type: bstring,
        src_ip: ip,
        src_port: port=(uint16),
        dest_ip: ip,
        dest_port: port=(uint16),
        vlan: [uint16],
        proto: bstring,
        app_proto: bstring,
        alert: {
          severity: uint16,
          signature: bstring,
          category: bstring,
          action: bstring,
          signature_id: uint64,
          gid: uint64,
          rev: uint64,
          metadata: {
            signature_severity: [bstring],
            former_category: [bstring],
            attack_target: [bstring],
            deployment: [bstring],
            affected_product: [bstring],
            created_at: [bstring],
            performance_impact: [bstring],
            updated_at: [bstring],
            malware_family: [bstring],
            tag: [bstring]
          }
        },
        flow_id: uint64,
        pcap_cnt: uint64,
        tx_id: uint64,
        icmp_code: uint64,
        icmp_type: uint64,
        tunnel: {
          src_ip: ip,
          src_port: port=(uint16),
          dest_ip: ip,
          dest_port: port=(uint16),
          proto: bstring,
          depth: uint64
        },
        community_id: bstring
      }
      filter event_type=alert | put . = shape(alert) | rename ts=timestamp
```

As would be expected, this fails if the `jq` binary isn't installed. I'm running on Linux in this case.

```
$ /opt/Brim/resources/app.asar.unpacked/zdeps/zed api new testpool
testpool: pool created

$ /opt/Brim/resources/app.asar.unpacked/zdeps/brimcap load -root "$HOME/.config/Brim/data/brimcap-root" -config suricata.yml -p testpool ~/wrccdc.pcap 
100.0% 500.0MB/500.0MB records=0 
Post "http://localhost:9867/pool/1s5nwutR6WQUyCWlV5oQynelwdi/log": suricata-wrapper.sh exited with code 127
stdout:
4/5/2021 -- 16:39:27 - <Notice> - This is Suricata version 6.0.2 RELEASE running in USER mode
4/5/2021 -- 16:39:27 - <Notice> - all 5 packet processing threads, 4 management threads initialized, engine started.
4/5/2021 -- 16:39:36 - <Notice> - Signal Received.  Stopping engine.
4/5/2021 -- 16:39:43 - <Notice> - Pcap-file module read 1 files, 1650919 packets, 473586644 bytes
stderr:
/home/phil/brimcap/examples/suricata-wrapper.sh: line 3: jq: command not found
```

However, this was initially tough to spot, because my original Brimcap YAML config also called out to a Zeek wrapper:

```
$ cat zeek-wrapper.sh 
#!/bin/bash
zeek -C -r - --exec "event zeek_init() { Log::disable_stream(PacketFilter::LOG); Log::disable_stream(LoadedScripts::LOG); }" local

$ diff zeek-suricata.yml suricata.yml 
2d1
<   - cmd: zeek-wrapper.sh
```

When I ran Brimcap using _that_ YAML, it effectively hangs without ever finishing, acting perhaps as if deadlocked. After waiting several minutes I tried to stop it via Control-C but it did not respond, so I then successfully stopped it via Control-\\. This does indeed produce a bunch of errors that mention lock/futex, which leads me to suspect deadlock.

```
$ /opt/Brim/resources/app.asar.unpacked/zdeps/zed api new testpool2
testpool2: pool created

$ /opt/Brim/resources/app.asar.unpacked/zdeps/brimcap load -root "$HOME/.config/Brim/data/brimcap-root" -config zeek-suricata.yml -p testpool2 ~/wrccdc.pcap 
100.0% 500.0MB/500.0MB records=392641 
^\SIGQUIT: quit
PC=0x4732e1 m=0 sigcode=128

goroutine 0 [idle]:
runtime.futex(0x14303f0, 0x80, 0x0, 0x0, 0x0, 0x7ffdbdde6982, 0x14302a0, 0x7ffdbdc4f7a8, 0x7ffdbdc4f7b8, 0x40dcbf, ...)
	/opt/hostedtoolcache/go/1.16.3/x64/src/runtime/sys_linux_amd64.s:579 +0x21
runtime.futexsleep(0x14303f0, 0x0, 0xffffffffffffffff)
	/opt/hostedtoolcache/go/1.16.3/x64/src/runtime/os_linux.go:44 +0x46
runtime.notesleep(0x14303f0)
	/opt/hostedtoolcache/go/1.16.3/x64/src/runtime/lock_futex.go:159 +0x9f
runtime.mPark()
	/opt/hostedtoolcache/go/1.16.3/x64/src/runtime/proc.go:1340 +0x39
runtime.stopm()
	/opt/hostedtoolcache/go/1.16.3/x64/src/runtime/proc.go:2257 +0x92
runtime.findrunnable(0xc000048800, 0x0)
	/opt/hostedtoolcache/go/1.16.3/x64/src/runtime/proc.go:2916 +0x72e
runtime.schedule()
	/opt/hostedtoolcache/go/1.16.3/x64/src/runtime/proc.go:3125 +0x2d7
runtime.park_m(0xc002042480)
	/opt/hostedtoolcache/go/1.16.3/x64/src/runtime/proc.go:3274 +0x9d
runtime.mcall(0x0)
	/opt/hostedtoolcache/go/1.16.3/x64/src/runtime/asm_amd64.s:327 +0x5b

goroutine 1 [semacquire]:
sync.runtime_Semacquire(0xc0020480f0)
	/opt/hostedtoolcache/go/1.16.3/x64/src/runtime/sema.go:56 +0x45
sync.(*WaitGroup).Wait(0xc0020480e8)
	/opt/hostedtoolcache/go/1.16.3/x64/src/sync/waitgroup.go:130 +0x65
github.com/brimdata/brimcap/ztail.(*Tailer).Close(0xc002048080, 0x0, 0xcd67a0)
	/home/runner/.npm/_cacache/tmp/git-clone-66762646/ztail/ztail.go:181 +0x65
github.com/brimdata/brimcap/analyzer.(*analyzer).Close(0xc0000342c0, 0x7fb0a2b4d008, 0xc0000342c0)
	/home/runner/.npm/_cacache/tmp/git-clone-66762646/analyzer/analyzer.go:167 +0xd3
github.com/brimdata/zed/zio.CloseReaders(0xc000368000, 0x2, 0x2, 0x0, 0x0)
	/home/runner/go/pkg/mod/github.com/brimdata/zed@v0.29.1-0.20210429222751-aa4170e0fff4/zio/zio.go:164 +0xd3
github.com/brimdata/brimcap/analyzer.(*combiner).Close(0xc00035e200, 0x0, 0x0)
	/home/runner/.npm/_cacache/tmp/git-clone-66762646/analyzer/combiner.go:82 +0xbd
github.com/brimdata/brimcap/cmd/brimcap/load.(*Command).Exec(0xc0002c81c0, 0xc000032110, 0x1, 0x1, 0xf6a680, 0xc0035e8150)
	/home/runner/.npm/_cacache/tmp/git-clone-66762646/cmd/brimcap/load/command.go:129 +0x871
github.com/brimdata/brimcap/cmd/brimcap/root.(*Command).Run(0xc0000811a0, 0xc000032110, 0x1, 0x1, 0x0, 0x0)
	/home/runner/.npm/_cacache/tmp/git-clone-66762646/cmd/brimcap/root/command.go:72 +0x1af
github.com/brimdata/zed/pkg/charm.path.run(0xc0002abf20, 0x2, 0x2, 0xc000032110, 0x1, 0x1, 0xc0002abf20, 0x2)
	/home/runner/go/pkg/mod/github.com/brimdata/zed@v0.29.1-0.20210429222751-aa4170e0fff4/pkg/charm/path.go:11 +0x8f
github.com/brimdata/zed/pkg/charm.(*Spec).ExecRoot(0x1423800, 0xc0000320a0, 0x8, 0x8, 0xffffffff, 0xc00008c058)
	/home/runner/go/pkg/mod/github.com/brimdata/zed@v0.29.1-0.20210429222751-aa4170e0fff4/pkg/charm/charm.go:63 +0x1e7
main.main()
	/home/runner/.npm/_cacache/tmp/git-clone-66762646/cmd/brimcap/main.go:18 +0x74

goroutine 50 [syscall, 1 minutes]:
os/signal.signal_recv(0x0)
	/opt/hostedtoolcache/go/1.16.3/x64/src/runtime/sigqueue.go:168 +0xa5
os/signal.loop()
	/opt/hostedtoolcache/go/1.16.3/x64/src/os/signal/signal_unix.go:23 +0x25
created by os/signal.Notify.func1.1
	/opt/hostedtoolcache/go/1.16.3/x64/src/os/signal/signal.go:151 +0x45

goroutine 51 [select, 1 minutes]:
github.com/brimdata/zed/pkg/signalctx.New.func1(0xc000080480, 0xc0002ac4b0, 0xc0002e0240, 0xf79cb8, 0xc00007e080)
	/home/runner/go/pkg/mod/github.com/brimdata/zed@v0.29.1-0.20210429222751-aa4170e0fff4/pkg/signalctx/signalctx.go:25 +0xad
created by github.com/brimdata/zed/pkg/signalctx.New
	/home/runner/go/pkg/mod/github.com/brimdata/zed@v0.29.1-0.20210429222751-aa4170e0fff4/pkg/signalctx/signalctx.go:24 +0x12f

goroutine 35 [select]:
github.com/brimdata/zed/pkg/display.(*Display).Run(0xc00016e000)
	/home/runner/go/pkg/mod/github.com/brimdata/zed@v0.29.1-0.20210429222751-aa4170e0fff4/pkg/display/display.go:50 +0xd4
created by github.com/brimdata/brimcap/cli/analyzecli.(*Display).Run
	/home/runner/.npm/_cacache/tmp/git-clone-66762646/cli/analyzecli/display.go:54 +0x145

goroutine 32 [chan send]:
github.com/brimdata/brimcap/ztail.(*Tailer).tailFile.func1(0xc002048080, 0xc0000440c0, 0xc003742f40, 0x1d)
	/home/runner/.npm/_cacache/tmp/git-clone-66762646/ztail/ztail.go:140 +0x22d
created by github.com/brimdata/brimcap/ztail.(*Tailer).tailFile
	/home/runner/.npm/_cacache/tmp/git-clone-66762646/ztail/ztail.go:120 +0x19b

goroutine 25 [semacquire]:
sync.runtime_Semacquire(0xc0020480f0)
	/opt/hostedtoolcache/go/1.16.3/x64/src/runtime/sema.go:56 +0x45
sync.(*WaitGroup).Wait(0xc0020480e8)
	/opt/hostedtoolcache/go/1.16.3/x64/src/sync/waitgroup.go:130 +0x65
github.com/brimdata/brimcap/ztail.(*Tailer).start(0xc002048080)
	/home/runner/.npm/_cacache/tmp/git-clone-66762646/ztail/ztail.go:86 +0x1aa
created by github.com/brimdata/brimcap/ztail.New
	/home/runner/.npm/_cacache/tmp/git-clone-66762646/ztail/ztail.go:49 +0x165

goroutine 16 [chan send]:
github.com/brimdata/brimcap/ztail.(*Tailer).tailFile.func1(0xc002048080, 0xc000044100, 0xc00003c3c0, 0x1e)
	/home/runner/.npm/_cacache/tmp/git-clone-66762646/ztail/ztail.go:140 +0x22d
created by github.com/brimdata/brimcap/ztail.(*Tailer).tailFile
	/home/runner/.npm/_cacache/tmp/git-clone-66762646/ztail/ztail.go:120 +0x19b

goroutine 60 [chan send]:
github.com/brimdata/brimcap/ztail.(*Tailer).tailFile.func1(0xc002048080, 0xc000368220, 0xc00003d200, 0x1f)
	/home/runner/.npm/_cacache/tmp/git-clone-66762646/ztail/ztail.go:140 +0x22d
created by github.com/brimdata/brimcap/ztail.(*Tailer).tailFile
	/home/runner/.npm/_cacache/tmp/git-clone-66762646/ztail/ztail.go:120 +0x19b

goroutine 30 [chan send]:
github.com/brimdata/brimcap/ztail.(*Tailer).tailFile.func1(0xc002048080, 0xc0020342e0, 0xc003041900, 0x1d)
	/home/runner/.npm/_cacache/tmp/git-clone-66762646/ztail/ztail.go:140 +0x22d
created by github.com/brimdata/brimcap/ztail.(*Tailer).tailFile
	/home/runner/.npm/_cacache/tmp/git-clone-66762646/ztail/ztail.go:120 +0x19b

goroutine 67 [chan send]:
github.com/brimdata/brimcap/ztail.(*Tailer).tailFile.func1(0xc002048080, 0xc0000449a0, 0xc003548500, 0x1d)
	/home/runner/.npm/_cacache/tmp/git-clone-66762646/ztail/ztail.go:140 +0x22d
created by github.com/brimdata/brimcap/ztail.(*Tailer).tailFile
	/home/runner/.npm/_cacache/tmp/git-clone-66762646/ztail/ztail.go:120 +0x19b

goroutine 82 [chan send]:
github.com/brimdata/brimcap/ztail.(*Tailer).tailFile.func1(0xc002048080, 0xc0030f5780, 0xc0020458c0, 0x1e)
	/home/runner/.npm/_cacache/tmp/git-clone-66762646/ztail/ztail.go:140 +0x22d
created by github.com/brimdata/brimcap/ztail.(*Tailer).tailFile
	/home/runner/.npm/_cacache/tmp/git-clone-66762646/ztail/ztail.go:120 +0x19b

rax    0xca
rbx    0x14302a0
rcx    0x4732e3
rdx    0x0
rdi    0x14303f0
rsi    0x80
rbp    0x7ffdbdc4f780
rsp    0x7ffdbdc4f738
r8     0x0
r9     0x0
r10    0x0
r11    0x286
r12    0x0
r13    0x26239
r14    0x80c000000000
r15    0x0
rip    0x4732e1
rflags 0x286
cs     0x33
fs     0x0
gs     0x0
```

**Comment by mattnibs** [comment]
We don't have a test for a process error with multiple analyzers. Will add this along with the fix.

**Comment by philrz** [comment]
I've been unable to manually verify this fix. In the original report, I could reliably reproduce the problem using `brimcap load`. However, per the verification notes for a more recent fix in [https://github.com/brimdata/brimcap/issues/112#issuecomment-875665763](https://github.com/brimdata/brimcap/issues/112#issuecomment-875665763), `brimcap load` effectively remains broken and is going to be phased out rather than fixed (#113). Meanwhile, I went back and tried to reproduce this issue using just `brimcap analyze` in hopes that would be a different baseline for manually verifying the fix, but I was unsuccessful. It always exited for me with an appropriate `jq: command not found` error message. I'm not sure what the different timing/logic might have been that made the problem so much easier to repro with `brimcap load`. However, @mattnibs confirmed that the original `brimcap load` repro steps led him to create the `brimcap analyze` test in [https://github.com/brimdata/brimcap/blob/main/cmd/brimcap/ztests/analyze-process-error.yaml](https://github.com/brimdata/brimcap/blob/main/cmd/brimcap/ztests/analyze-process-error.yaml) as part of the linked PR #110 and that he'd used that to reliably reproduce the deadlock before the fix and that it runs reliably now that we have the fix. Therefore it looks like that'll have to stand as verification, and brimdata/brim#1716 is what'll get Brimcap working for Brim users again.

Thanks @mattnibs!
